### DISCUSSION TITLE
My next.js app goes blank after deploying on cPanel

### DISCUSSION THREAD

**Discussion by Wiseard** [vercel/next.js#54728] [Help]
### Summary

I deployed my Next.js app on cPanel and its blank. It works fine locally. Here are the steps I followed :

1.First, I created a custom server.js in my root directory with the following code :

`const { createServer } = require('http')
const { parse } = require('url')
const next = require('next')

const dev = process.env.NODE_ENV !== 'production'
const hostname = 'localhost'
const port = 3000
// when using middleware `hostname` and `port` must be provided below
const app = next({ dev, hostname, port })
const handle = app.getRequestHandler()

app.prepare().then(() => {
  createServer(async (req, res) => {
    try {
      // Be sure to pass `true` as the second argument to `url.parse`.
      // This tells it to parse the query portion of the URL.
      const parsedUrl = parse(req.url, true)
      const { pathname, query } = parsedUrl

      if (pathname === '/a') {
        await app.render(req, res, '/a', query)
      } else if (pathname === '/b') {
        await app.render(req, res, '/b', query)
      } else {
        await handle(req, res, parsedUrl)
      }
    } catch (err) {
      console.error('Error occurred handling', req.url, err)
      res.statusCode = 500
      res.end('internal server error')
    }
  })
    .once('error', (err) => {
      console.error(err)
      process.exit(1)
    })
    .listen(port, () => {
      console.log(`> Ready on http://${hostname}:${port}`)
    })
})
`

2. Then I edited the package.json file to set the environment to production on the start script :

` "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "NODE_ENV=production node server.js",
    "lint": "next lint"
  },`

3. npm run build and compress the following files :

<img width="587" alt="build" src="https://github.com/vercel/next.js/assets/122892355/850dde4e-e866-4cd7-b12c-c9f4f98e578f">

4. Imported and extracted all in the File Manager Root Directory :

<img width="909" alt="file-manager" src="https://github.com/vercel/next.js/assets/122892355/f8cb303d-356a-453d-820d-e8c292afcf81">

5. Created my application on Setup Node.js App with the following information :

<img width="935" alt="node-app" src="https://github.com/vercel/next.js/assets/122892355/2ee5c2cc-7eeb-4d8f-94ad-0979cad99454">

6. Stop Application - Run NPM install and then restart the application

After all these steps, my app is blank, and I don't know what I did wrong.

Here is a [Github link](https://github.com/Wiseard/nextjs-app) where you can see all the app code. I'm stuck since yesterday and I appreciate any help you can give me to resolve this problem. I know the issue is due to my app but I can't figure it out. But I'm pretty sure the main problem is located on the server.js file.

I hope all the information I provided could help you to understand.

**Comment by rahulsharmacode** [comment]
I am also facing same issue, my app also works fine in local but, after hosting in cpanel with all steps followed, the app goes blank.

**Comment by yiaoma** [comment]
Same here, we had a website hosted on cpanel for almost a year already, it crashed when I updated my packages to match nextjs13

**Reply by harisahmedz** [reply]
did u find any solution?

**Comment by nicolez9911** [comment]
Same thing. I tested some versions, 13.4.13 is the last one that works for me. With later versions I just get empty responses (even with the example project).

**Reply by yiaoma** [reply]
How do you downgrade nextjs to 13.4.13, do you just put this version in package.json file?

**Reply by nicolez9911** [reply]
Yes, that‘s what I did. And then of course `npm install` and a new build. I‘m nor sure if that‘s the way to go, but it worked for me..

**Reply by yiaoma** [reply]
Thank you, it did work but only on nodesj application. We were using passenger before, it is completely broken right now even with downgraded version. Have you had any issues with next images ? I'm getting current error: ""url" parameter is valid but upstream response is invalid"

**Reply by nicolez9911** [reply]
didn't have any issues with images..

**Reply by NafeesIS** [reply]
The process worked to end to end, but when I open the URL, I get the notification "It works! NodeJS 20.18.3", Would you know why?

**Comment by comtrast-remi** [comment]
Same issue here but on a Plesk server. I've also made changes to the `package.json` file and created a `server.js` file. 

I'm using Next 13.4.19 and Node 16.20.

**Reply by yiaoma** [reply]
At the moment the only version we've managed to run without any problems is 13.4.4

**Comment by one-and-only** [comment]
Bumping this, as the issue still persists with Next.JS 14 and 15-RC.0.

**Reply by ByAnthony** [reply]
Managed to deploy on CPanel with Next.js 14.2.4. Wrote a [wee guide](https://github.com/ByAnthony/newzealandtunnellers/blob/main/docs/deployment.md); hope that helps 🙏

**Reply by one-and-only** [reply]
> Managed to deploy on CPanel with Next.js 14.2.4. Wrote a [wee guide](https://github.com/ByAnthony/newzealandtunnellers/blob/main/docs/deployment.md); hope that helps 🙏

I appreciate the effort. However, the issue here seems that after a while, all page responses end up return literally an empty HTML body after all the headers. The server runs for the first few hours or so, but it eventually "dies". The issue still persists in its original form (I just testedd with latest Canary on Node 22).

**Reply by NafeesIS** [reply]
The process worked to end to end, but when I open the URL, I get the notification "It works! NodeJS 20.18.3", Would you know why?

**Comment by NafeesIS** [comment]
The process worked to end to end, but when I open the URL, I get the notification "It works! NodeJS 20.18.3", Would you know why?

**Reply by designdrumm** [reply]
Take a look inside app.js in the parent directory. More than likely your node instance start page is set to this js file.

**Comment by designdrumm** [comment]
Next.js needs node v18.17..  min. I believe to work on a server.

**Comment by designdrumm** [comment]
I am having the same issue. 
I did notice this code in you post.

```
if (pathname === '/a') {
    await app.render(req, res, '/a', query)
  } else if (pathname === '/b') {
    await app.render(req, res, '/b', query)
  } else {
    await handle(req, res, parsedUrl)
  }
```

Fix /a and /b, they should be pages related to your app.
Essentially, I think they are giving you the ability to use queries here for 
only pages that require them otherwise leaving them out.

So for example if you have a cart page and say a blog page that take url queries, 
then they are /a and /b . (pathname === '/cart' or pathname === '/blog') 
otherwise all other pages just load via the url.

Not sure if this will fix your problem though. 
I am still in the middle of testing. Why I found this discussion!

**Comment by alnetarg** [comment]
Hi team,

I’m following up on the issue regarding deploying a simple Next.js application on cPanel. Despite allocating 4GB of RAM, I’m unable to get even a basic test project running.

I created the required server.js file as per documentation:

import express from "express";
import next from "next";

const port = process.env.PORT || 3000;
const dev = process.env.NODE_ENV !== "production";
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
    const server = express();

    server.all("*", (req, res) => {
        return handle(req, res);
    });

    server.listen(port, (err) => {
        if (err) throw err;
        console.log(`🚀 Server running at http://localhost:${port}`);
    });
});


My package.json contains the following script:

"scripts": {
  "dev": "next dev --turbopack",
  "build": "next build",
  "start": "NODE_ENV=production node server.js",
  "lint": "next lint"
}


Node.js version: 20.17.0
Application mode: Production

However, when running the start script through the UI using "RUN JS SCRIPT", I get this error:

RangeError: WebAssembly.instantiate(): Out of memory: Cannot allocate Wasm memory for new instance
    at lazyllhttp (node:internal/deps/undici/undici:5560:32)


The Node.js process fails with a WebAssembly memory allocation error, although the application only consumes ~24MB of RAM when tested on a Debian VM with 4GB RAM (same conditions, runs fine).

Limits shown include:

Max resident set: 4GB

Max address space: 4GB

I’ve already reviewed that the issue is not on the application side. Could you please:

Confirm if this may be due to LVE limits or a deeper issue with the platform's WebAssembly/node runtime?

Attempt to deploy a test Next.js app yourselves on this environment for validation (perhaps using the example starter project)?

Let me know if any tuning is needed for this kind of app.

Thanks in advance for the support. Let me know if you need access logs or configs.

Best regards,

**Comment by designdrumm** [comment]
I was told by my hosting tech staff that the memory error is because of the Node setup UI and memory limits can not be increased for this in CPanel. They advised me to use terminal for this instead.

Also. I found this regarding your code you posted.
Specifically the commented text.

    ```
    ...

    app.prepare().then(() => {
      const server = express();

      // Custom server-side logic or API routes can go here
      // For example:
      // server.get('/custom-route', (req, res) => {
      //   res.send('Hello from custom route!');
      // });

      // Handle all other requests with Next.js
      server.all('*', (req, res) => {
        return handle(req, res);
      });

      server.listen(port, (err) => {
        if (err) throw err;
        console.log(`> Ready on http://localhost:${port}`);
      });
    });
```

**Comment by designdrumm** [comment]
I have had some success with loading my app. 
When I set the application start file to the server.js in the Node Setup UI and replace the "NODE_ENV=production node server.js" text from my package.json file to "next start". I then start my app and "run script" selecting start, I get my site html loading with broken links and mostly just the raw html exposed from my front page. But my app loaded! Again though, I am getting the memory error in the UI. My suspicion is that I need to take the techs advice and try with terminal however, I am not savvy with terminal.

**Comment by designdrumm** [comment]
Hello All,
Not sure if anyone needs this at this point, but I found a nifty way to gather up the files needed to upload to the server. It's a clean low footprint way. Only gathers the necessary files needed to run your node app.  

For context I am building a Next.js app and using Node Setup in CPanel. 

One note, once you upload or before you upload the files it creates (deploy folder contents), be sure to delete the postbuild directive in the generated deploy folder package.json. The postbuild is only intended for local builds. If used on the server in the node setup, you might create files on your server and not know where they are stored. I personally, manually upload the files it creates to the appropriate folders to avoid this. Once the postbuild is removed, you can successfully run NPM in the Node setup as expected and not have rogue files on your server.

HTH

(inside package.json):

`
"scripts": {
    "dev": "next dev",
    "build": "npx next build",
    "postbuild": "mkdirp .next/standalone/.next/static && cp -R public .next/standalone/public && cp -R .next/static/. .next/standalone/.next/static/ && cpy .env .next/standalone && rm -rf deploy && mkdir -p deploy && cp -R .next/standalone/. deploy && cpy next.config.js tsconfig.json next-env.d.ts package-lock.json babelrc .eslintrc.json deploy && mkdir -p deploy/.next && cp -R .next/static deploy/.next/ && cp -R public deploy/public && ([ -f .htaccess ] && cp .htaccess deploy/prod.htaccess || true)",
    "start": "node server.js",
    "lint": "next lint"
  },

`
  This is what that postbuild produces for me.
  deploy/
  —  package.json (remove postbuild from here)
  —  prod.htaccess (remove prod from file name if using on server)
  —  public/
  —  pages/
  —  .env
  —  .next/
  —  node_modules/
  —  server.js
  —  next.config.js
  —  package-lock.json
  —  tsconfig.json
  —  next-env.d.ts
  —  babelrc
  —  .eslintrc.json
  
  You can change the files it copies by modifying the postbuild directive to taste. The .htaccess intentionally has the prod attached to the front of it so you don't overwrite yours on your server accidentally. Make sure you remove the prod before running your app or don't include copying the .htaccess file in the postbuild. You can modify this postbuild however you need! That's just what I came up with for my setup.
  
 My server.js:

  `
import { createServer } from "http";
import next from "next";

const port = parseInt(process.env.PORT || "3000", 10);
const dev = process.env.NODE_ENV !== "production";
const app = next({ dev });
const handle = app.getRequestHandler();

app.prepare().then(() => {
  createServer((req, res) => {
    const protocol = req.headers["x-forwarded-proto"] || "https";
    const parsedUrl = new URL(req.url, `${protocol}://${req.headers.host}`);
    handle(req, res, parsedUrl);
  }).listen(port, () => {
    console.log(
      `> Server listening on port ${port} as ${
        dev ? "development" : "production"
      }`
    );
  });
});
`
The created deploy folder will show up inside your project folder where you build from.

One other note that may be pertinent is that I am on a Mac using VSCode and that uses Terminal. 
If you are on a windows computer, this might not work as written. If so, you can probably paste the postbuild code in Google AI mode and ask for the alternative that'll work with your system.
